src/auth/mysql.rs

```diff
@@ -0,0 +1,362 @@
+#![allow(unused_qualifications)]
+//! MySql authenticator module
+use argon2rs;
+
+use diesel::prelude::*;
+use diesel::mysql::MysqlConnection;
+use diesel::result::ConnectionError;
+
+// FIXME: Remove dependency on `ring`.
+use ring::test;
+use ring::constant_time::verify_slices_are_equal;
+
+use {Error, JsonMap, JsonValue};
+use super::{AuthenticationResult, Basic};
+
+/// Error mapping for `ConnectionError`
+impl From<ConnectionError> for Error {
+    fn from(e: ConnectionError) -> Error {
+        Error::GenericError(e.to_string())
+    }
+}
+
+// Code for conversion to hex stolen from rustc-serialize:
+// https://doc.rust-lang.org/rustc-serialize/src/rustc_serialize/hex.rs.html
+
+/// MySql user record
+#[derive(Queryable)]
+pub struct User {
+    username: String,
+    pw_hash: String,
+    salt: String,
+}
+
+/// A simple authenticator that uses a MySql backed user database.
+///
+/// Requires the `mysql_authenticator` feature, which is enabled by default.
+///
+/// The user database should be a MySql database with a table of the following format:
+/// username(VARCHAR(255)), pw_hash(VARCHAR(255)), salt(VARCHAR(255))
+///
+/// # Password Hashing
+/// See `MySqlAuthenticator::hash_password` for the implementation of password hashing.
+/// The password is hashed using the [`argon2i`](https://github.com/p-h-c/phc-winner-argon2) algorithm with
+/// a randomly generated salt.
+pub struct MySqlAuthenticator {
+    database_uri: String
```

Comment:
`MysqlConnection` is not `Send + Sync`, but we should not really be reconnecting every time we want to make a query.

Instead, we should use [`r2d2-diesel`](https://github.com/diesel-rs/r2d2-diesel) to hold a connection pool for MySQL which is created during initialisation. Then the `MySqlAuthenticator` struct can hold a `ConnectionManager<MysqlConnection>` instead.

I would be happy to accept this PR without implementing this change and then raise an issue for this to be resolved in future PRs.